### PR TITLE
dropbear: add option to generate ECDSA host key

### PR DESCRIPTION
**Pull request by sch-m**
Signed-off-by: Martin Schiller <ms@dev.tdt.de>

### CODE REVIEW

File: package/network/services/dropbear/Makefile [hunk 1]
```diff
@@ -142,6 +144,12 @@ define Package/dropbear/install
 	$(INSTALL_DIR) $(1)/usr/lib/opkg/info
 	$(INSTALL_DIR) $(1)/etc/dropbear
 	touch $(1)/etc/dropbear/dropbear_rsa_host_key
+ifeq ($(CONFIG_DROPBEAR_ECDSA_KEYGEN),y)
+	sed -i \
+		'/[ -s \/etc\/dropbear\/dropbear_rsa_host_key ] || keygen rsa/a \\t[ -s \/etc\/dropbear\/dropbear_ecdsa_host_key ] || keygen ecdsa' \
+		$(1)/etc/init.d/dropbear
+	touch $(1)/etc/dropbear/dropbear_ecdsa_host_key
+endif
```
Comment (blogic) [inline package/network/services/dropbear/Makefile]: this is way too ugly. can't you runtime detect what key should be generated ?

**Comment by rockdrilla** [comment]
Hi! Have you tested #915 ?

File: package/network/services/dropbear/Makefile (hunk 1 above)
Comment (sch-m) [reply package/network/services/dropbear/Makefile]: The "dropbearkey -h" output could maybe used to get the supported keytypes.

I thought there are maybe users who have ECC support activated, but don't want to generate an ecdsa host key. So I deciced to add an new extra option. 
Autodetection would be difficult in this case.

I don't know if this is really a practical usecase and I also considered to simply add ecdsa keygen if DROPBEAR_ECC is activated.

**Comment by sch-m** [comment]
@rockdrilla Your keytest() and keygen() functions in #915 will always try to generate an ecdsa key, but that won't work if neither DROPBEAR_ECC or DROPBEAR_ECC_FULL is activated. Or did I miss something?

**Comment by rockdrilla** [comment]
@sch-m, i apologise that my code looks non-linear, but i'll try to describe it here.

(dropbear) package ships empty host key files - rsa (always) and ecdsa (only if DROPBEAR_ECC is selected).

later, init script tests these files during start_service() by calling keytest():
- if host key file doesn't exist, then key type is skipped (no error);
- if host key isn't empty, then it's considered as "good";
- otherwise, error is returned.

now, if keytest() returned error, init script calls keygen() which does pretty same tests against files: if host key file is empty, then key generation is called for selected key type.

ps: config option DROPBEAR_ECC_FULL depends on DROPBEAR_ECC.

**Comment by sch-m** [comment]
@rockdrilla Ah ok, but this will change the existing behavior. At the moment, the host key is also generated when the host key file doesn't exist.

**Comment by rockdrilla** [comment]
@sch-m, oh... my bad, i'll rewrite keytest/keygen login and modify own pull request.

**Comment by sch-m** [comment]
OK.

**Comment by rockdrilla** [comment]
@sch-m, done. Please review cf2ef0c904.

**Comment by sch-m** [comment]
@rockdrilla LGTM

**Comment by ynezz** [comment]
Closing as this should be probably handled in #915 and 2019.77 version bump.

**Pull request closed without merging**